File: Bowdoin Dining/Menus.m
//
//  Menus.m
//  Bowdoin Dining
//
//  Created by Ruben on 7/11/14.
//
//

#import "Menus.h"
#import "GDataXMLNode.h"
#import "BowdoinDining-Swift.h"
@interface Menus ()


@end

NSString *serverURL = @"http://www.bowdoin.edu/atreus/lib/xml/";
@implementation Menus

//format an NSDate for our use
+ (NSMutableArray *)formatDate: (NSDate *) todayDate {
    //load gregorian calendar
    NSCalendar *calendar = [[NSCalendar alloc] initWithCalendarIdentifier:NSCalendarIdentifierGregorian];
    [calendar setLocale:[[NSLocale alloc] initWithLocaleIdentifier:@"en-US"]];
    
    //create DateComponents from the NSDate and NSCalendar
    NSDateComponents *today = [calendar components:NSCalendarUnitYear | NSCalendarUnitWeekOfYear | NSCalendarUnitWeekday fromDate:todayDate];
    
    //calculate offset from sunday (day menu begins)
    NSInteger offset  = [today weekday];
    
    //set current day to sunday (first day) of this week, and create NSDateComponents for that day
    [today setWeekday:1]; //Sunday
    NSDate *lastSundayDate = [calendar dateFromComponents:today];
    NSDateComponents *lastSunday = [calendar components:NSCalendarUnitDay | NSCalendarUnitYear | NSCalendarUnitMonth | NSCalendarUnitWeekday fromDate:lastSundayDate];
    
    //store info about last sunday's date for use with Bowdoin XML API
    NSInteger day     = [lastSunday day];
    NSInteger month   = [lastSunday month]-1;
    NSInteger year    = [lastSunday year];
    
    //return info in array
    NSMutableArray *results  = [[NSMutableArray alloc] init];
    [results addObject: [NSNumber numberWithLong:day]];
    [results addObject: [NSNumber numberWithLong:month]];
    [results addObject: [NSNumber numberWithLong:year]];
    [results addObject: [NSNumber numberWithLong:offset]];
    return results;
}

//create a menu url for a given day
+ (NSString *)getMenuUrlForDay: (NSInteger) day Month: (NSInteger) month Year: (NSInteger) year Offset: (NSInteger) offset {
    NSString* url = [[[[serverURL
                stringByAppendingString: [NSString stringWithFormat:@"%ld",  (long)year]]
                stringByAppendingString: [NSString stringWithFormat:@"-%ld", (long)month]]
                stringByAppendingString: [NSString stringWithFormat:@"-%ld", (long)day]]
                stringByAppendingString: [NSString stringWithFormat:@"/%ld.xml", (long)offset]];
    return url;
}

//create a local path for a given day
+ (NSString *)getLocalPathForDay: (NSInteger) day Month: (NSInteger) month Year: (NSInteger) year Offset: (NSInteger) offset {
    NSString* local = [[[[@"local"
                stringByAppendingString: [NSString stringWithFormat:@"-%ld",  (long)year]]
                stringByAppendingString: [NSString stringWithFormat:@"-%ld", (long)month]]
                stringByAppendingString: [NSString stringWithFormat:@"-%ld", (long)day]]
                stringByAppendingString: [NSString stringWithFormat:@"-%ld.xml", (long)offset]];
    
    NSArray* paths = NSSearchPathForDirectoriesInDomains(NSCachesDirectory, NSUserDomainMask, YES);
    NSString* path = [[paths firstObject] stringByAppendingPathComponent:local];
    return path;
}

//load menu for a given day
+ (NSData *)loadMenuForDay: (NSInteger) day Month: (NSInteger) month Year: (NSInteger) year Offset: (NSInteger) offset {
    //first, search local path in case cached
    NSString* path = [self getLocalPathForDay: day Month: month Year: year Offset: offset];
    BOOL fileExists = [[NSFileManager defaultManager] fileExistsAtPath:path];
    if(fileExists) {
        //if cached, return cached file
        NSData *cached = [NSData dataWithContentsOfFile: path];
        return cached;
    } else { //else not cached
        //begin network activity
        [[UIApplication sharedApplication] setNetworkActivityIndicatorVisible:YES];
        
        //download menu for this day
        NSString *urlString = [self getMenuUrlForDay: day Month: month Year: year Offset: offset];
        NSURL *url = [[NSURL alloc] initWithString: urlString];
        NSError *error = nil;
        NSData *xmlData = [NSMutableData dataWithContentsOfURL: url options: 0 error: &error];
        if(error) NSLog(@"%@", [error debugDescription]);

        //end network activity
        [[UIApplication sharedApplication] setNetworkActivityIndicatorVisible:NO];
        
        //cache file
        [xmlData writeToFile:path atomically:YES];
        
        //return downloaded file
        return xmlData;
    }
}

//creates a Menu (NSMutableArray*) from an NSData* XML file for a given meal/location and filters
+ (NSMutableArray *)createMenuFromXML:(NSData *) xmlData forMeal: (NSInteger) mealId onWeekday: (BOOL) weekday atLocation: (NSInteger) locationId withFilters: (NSMutableArray *) filters {
    NSError *error;
    //Create Google XML parsing object from NSData, grab "<meal>"s below root
    GDataXMLDocument *doc = [[GDataXMLDocument alloc] initWithData:xmlData options:0 error:&error];
    GDataXMLElement *root = doc.rootElement;
    NSArray *meals = [root elementsForName:@"meal"];
    
    //compensates for disappearing meals
    switch (mealId) {
        case 0:
            if(!weekday) {
                mealId = 1;
            }
            break;
        case 1:
            if(weekday) {
                mealId = 2;
            } else {
                mealId = 3;
            }
            break;
        case 2:
            mealId = 3;
        default:
            break;
    }

    GDataXMLElement *meal = [meals objectAtIndex: mealId];
    
    //each meal has two units (locations), create an XML Element for this locationId's menu
    NSArray *units = [meal elementsForName:@"unit"];
    GDataXMLElement *unit = [units objectAtIndex: locationId];
    GDataXMLElement *menu = [[unit elementsForName: @"menu"] firstObject];
    
    //create array for records (menu items), initialize array of courses (a menu item attribute)
    NSArray *menuItems = [menu elementsForName: @"record"];
    NSMutableArray *courses = [[NSMutableArray alloc] init];
    
    //if there are menu items available, loop through them
    if(menuItems.count > 0) {
        for(GDataXMLElement *item in menuItems) {
            //determine course for this item and check if it already exists in our courses array
            GDataXMLElement *courseObject = (GDataXMLElement *) [[item elementsForName: @"course"] firstObject];
            NSInteger coursePosition = -1;
            for(int i = 0; i < courses.count; i++) {
                Course *course = courses[i];
                if([course.courseName isEqualToString: courseObject.stringValue])
                    coursePosition = i;
            }
            
            //grab information about this menu item: name & id
            GDataXMLElement *item_name = [[item elementsForName:@"webLongName"] firstObject];
            GDataXMLElement *item_id = [[item elementsForName:@"itemID"] firstObject];
            
            //create regex for removing diet attributes from item name, find matches in string
            NSRegularExpression *regex = [NSRegularExpression regularExpressionWithPattern:@"\\b(GF|VE|V|L)\\b" options:0 error:&error];
            NSArray *details  = [regex matchesInString:item_name.stringValue options:0 range:NSMakeRange(0, [item_name.stringValue length])];
            
            //store returned attributes into string for presentation
            NSString *detail = @"";
            if(details.count) { //if there were matches, loop through them and add them to string
                for(int i = 0; i < details.count; i++) {
                    NSTextCheckingResult *special = (NSTextCheckingResult *) [details objectAtIndex:i];
                    detail = [[detail stringByAppendingString: [[item_name.stringValue
                                substringWithRange: special.range]
                                stringByTrimmingCharactersInSet: [NSCharacterSet whitespaceCharacterSet]]]
                                stringByAppendingString:@" "];
                }
            }
            
            //replace matches with empty space
            NSString *cleaned = [[[regex stringByReplacingMatchesInString:item_name.stringValue
                                        options:0
                                        range:NSMakeRange(0, [item_name.stringValue length]) withTemplate:@""]
                                        stringByReplacingOccurrencesOfString:@"(" withString:@""]
                                        stringByReplacingOccurrencesOfString:@")" withString:@""];
            
            //break up string of diet attributes in array for filtering
            NSArray *attributes = [detail componentsSeparatedByString:@" "];
            
            //check if diet attributes match our filter
            NSMutableSet *overlap = [NSMutableSet setWithArray:filters];
            [overlap intersectSet:[NSSet setWithArray:attributes]];
            
            //if there is no active filter, or this item passes our filter
            if(!filters.count || [overlap allObjects].count) {
                //declare this course
                Course *thiscourse;
                
                //if course already exists in our array
                if(coursePosition >= 0) {
                    //grab a copy of it, and and add this item to the course
                    thiscourse = courses[coursePosition];
                    
                    MenuItem *item = [[MenuItem alloc] init];
                    item.name = cleaned;
                    item.itemId = item_id.stringValue;
                    item.descriptors = detail;
                    
                    [thiscourse.menuItems addObject:item];
                } else { //new course, create it and add item to it
                    thiscourse = [[Course alloc] init];
                    thiscourse.courseName = courseObject.stringValue;
                    
                    MenuItem *item = [[MenuItem alloc] init];
                    item.name = cleaned;
                    item.itemId = item_id.stringValue;
                    item.descriptors = detail;
                    
                    [thiscourse.menuItems addObject:item];
                    [courses addObject: thiscourse];
                }
            }
        }
    }
    //no menu items available, add error item to courses array
    else {
        Course *closed = [[Course alloc] init];
        closed.courseName = @"";
        MenuItem *item = [[MenuItem alloc] init];
        item.name = @"No Menu Available";
        item.itemId = @"NA";
        [closed.menuItems addObject: item];
        [courses addObject: closed];
    }
    return courses; //return array of courses
}

@end

File: Bowdoin Dining/Menus.swift
//
//  Menus.swift
//  Bowdoin Dining
//
//  Created by Ruben Martinez Jr on 7/25/14.
//
//

import Foundation
import UIKit

class Menus {
    static let menuUrl = "https://apps.bowdoin.edu/orestes/api.jsp"
    static let blackList = ["Salad Bar", "...", "Salads", "Salad Bar L & D -- Summer", "Deli Bar"]
    static let unitBase = 48
    
    class func cachePath() -> String {
        let cachePath = URL(fileURLWithPath: NSSearchPathForDirectoriesInDomains(FileManager.SearchPathDirectory.cachesDirectory, FileManager.SearchPathDomainMask.userDomainMask, true).first!)
        let menusPath = cachePath.appendingPathComponent("menus")
        
        //create path if it doesn't exist
        if (!FileManager.default.fileExists(atPath: menusPath.path)) {
            do {
                try FileManager.default.createDirectory(atPath: menusPath.path, withIntermediateDirectories: false, attributes: nil)
            } catch {
                print("Cache Path Error")
            }
        }
        
        return menusPath.path
    }
    
    class func clearOldCache() {
        let cachePath = self.cachePath()
        do {
            let allCache = try FileManager.default.contentsOfDirectory(atPath: cachePath)

            for item in allCache {
                let today = Date()
                let attributes = try FileManager.default.attributesOfItem(atPath: item)
                let creationDate = attributes[FileAttributeKey.creationDate] as! Date
                if creationDate.compare(today) == ComparisonResult.orderedAscending {
                    try FileManager.default.removeItem(atPath: item)
                }
            }
        } catch {
            // Unexpected error!
            print("Could not clear cache.")
        }
    }
    
    class func apiMenuURL(day: Int, month: Int, year: Int, unit: Int, mealName: String) -> String {
        let monthString = "\(month)".leftPadding(toLength: 2, withPad: "0")
        let dayString = "\(day)".leftPadding(toLength: 2, withPad: "0")
        let locationUnit = Menus.unitBase + unit
        
        return "\(Menus.menuUrl)?date=\(year)\(monthString)\(dayString)&unit=\(locationUnit)&meal=\(mealName)"
    }
    
    class func cacheMenuURL(day: Int, month: Int, year : Int, unit: Int, mealName: String) -> String {
        let monthString = NSString(string: "\(month)").padding(toLength: 2, withPad: "0", startingAt: 0)
        let dayString = NSString(string: "\(day)").padding(toLength: 2, withPad: "0", startingAt: 0)
        
        let file = "local-\(year)-\(monthString)-\(dayString)-\(unit)-\(mealName).xml"
        return "\(self.cachePath())\(file)"
    }
    
    class func writeMenuToCache(day : Int, month : Int, year : Int, unit: Int, mealName: String, data: Data?) {
        if let menu = data {
            let path = Menus.cacheMenuURL(day: day, month: month, year: year, unit: unit, mealName: mealName)
            let url = URL(string: path)!
            
            //cache file, return
            try? menu.write(to: url, options: [.atomic])
        }
    }
    
    class func readMenuFromCache(day: Int, month: Int, year: Int, unit: Int, mealName: String) -> Data? {
        let path = self.cacheMenuURL(day: day, month: month, year: year, unit: unit, mealName: mealName)
        
        return FileManager.default.fileExists(atPath: path) as Bool
            ? try? Data(contentsOf: URL(fileURLWithPath: path))
            : nil;
    }
    
    class func readMenuFromAPI(day : Int, month : Int, year : Int, unit: Int, mealName: String) -> Data? {
        let path = self.apiMenuURL(day: day, month: month, year: year, unit: unit, mealName: mealName)
        let url = URL(string: path)!
        
        // Begin network activity.
        Menus.activityIndicator(on: true)
        
        // Download menu for this day.
        let xmlData = try? Data(contentsOf: url)
        
        // End network activity.
        Menus.activityIndicator(on: false)
        
        // Cache data.
        Menus.writeMenuToCache(day: day, month: month, year: year, unit: unit, mealName: mealName, data: xmlData)
        
        return xmlData as Data?
    }
    
    //load menu for a given day
    class func loadMenu(date: Date, unit: Int, meal: Int) -> Data? {
        // Extract day/month/year.
        let components      = date.getDayMonthYear()
        let day             = components.day
        let month           = components.month
        let year            = components.year
        let mealName        = Menus.getMealNameForMeal(date: date, meal: meal)
        
        // Default to cached menu if it exists.
        var menu = Menus.readMenuFromCache(day: day, month: month, year: year, unit: unit, mealName: mealName)
        
        // Menu not cached, read externally.
        if menu == nil {
            menu = Menus.readMenuFromAPI(day: day, month: month, year: year, unit: unit, mealName: mealName)
        }
        
        return menu
    }
    
    class func createMenuFromXML(_ xmlData : Data, withFilters filters : [String]) -> [Course] {
        // Read the document.
        let doc = try! GDataXMLDocument(data: xmlData, options: 0)
        
        // Check if we have a menu & menu items.
        let menu = doc.rootElement()
        let menuItems = menu?.elements(forName: "record") as? [GDataXMLElement] ?? []
        
        // If no menu array, give up here.
        if menuItems.count == 0 {
            return Menus.errorMenu()
        }
        
        var courses : [Course] = []
        
        for item in menuItems {
            // Grab information about this menu item: course, name & id.
            let courseName = (item.elements(forName: "course") as? [GDataXMLElement])?.first?.stringValue() ?? ""
            let itemName = (item.elements(forName: "webLongName") as? [GDataXMLElement])?.first?.stringValue() ?? ""
            let itemId   = (item.elements(forName: "itemID") as? [GDataXMLElement])?.first?.stringValue() ?? ""
            
            // Give up here if we're ignoring this item.
            let passesBlackList = !Menus.blackList.contains(itemName)
            if !passesBlackList {
                continue;
            }
            
            // Parse out attributes (e.g. GF)
            let (name, attributes) = Menus.splitAttributesFromItemName(itemName)
            
            // Give up here if this item doesn't pass our filter.
            let passesFilter = filters.count == 0 || attributes.intersects(filters)
            if !passesFilter {
                continue;
            }
            
            // Check if course already exists in our courses array
            let existingCourse = courses.first(where: { (item) -> Bool in
                item.courseName == courseName
            });
            let courseExists = existingCourse != nil
            
            // Use existing or create new course.
            let course = courseExists ? existingCourse! : Course(name: courseName)
            
            // Create a new menuItem.
            let item = MenuItem(name: name, itemId: itemId, attributes: attributes)
            
            // Add to course menu items.
            course.menuItems.append(item)
            
            //if did not exist, add it to our courses.
            if !courseExists {
                courses.append(course)
            }
        }
        
        return courses
    }
    
    class func splitAttributesFromItemName(_ itemName: String) -> (name: String, attributes: [String]) {
        // Create regex for removing diet attributes from item name, find matches in string
        let regex = try! NSRegularExpression(pattern: "\\b(NGI|GF|DF|VE|V|L|H)\\b", options: [])
        
        let attributeMatches = regex.matches(in: itemName, options: [], range: NSMakeRange(0, itemName.count)) as NSArray
        
        // Store diet attributes for filtering
        var attributes : [String] = []
        
        // If there were matches, loop through them and add them to string
        if attributeMatches.count > 0 {
            for i in 0 ..< attributeMatches.count {
                let special = attributeMatches.object(at: i) as! NSTextCheckingResult
                attributes += [(itemName as NSString).substring(with: special.range) as String]
            }
        }
        
        //replace matches with empty space
        let name = regex.stringByReplacingMatches(in: itemName, options: [], range: NSMakeRange(0, itemName.count), withTemplate: "").replacingOccurrences(of: "(", with: "").replacingOccurrences(of: ")", with: "").trim()
        
        return (
            name,
            attributes
        )
    }
    
    class func errorMenu() -> [Course] {
        //no menu items available, add error item to courses array
        let closed = Course(name: "")
        
        let item = MenuItem(name: "No Menu Available", itemId: "NA", attributes: nil)
        closed.menuItems.append(item)
        
        return [closed]
    }
    
    class func loadFavoritesDataForCourses(_ courses : [Course]) -> [String : Int] {
        var itemIds : [String] = []
        for course in courses {
            for item in course.menuItems {
                if item.itemId != "NA" {
                    itemIds.append(item.itemId)
                }
            }
        }
        
        //begin network activity
        Menus.activityIndicator(on: true)
        
        //make request to get info from itemIds
        let endPointURL = "https://app.bowdoin.menu/favorites/\(itemIds.combine(","))"
        let url = URL(string: endPointURL)
        let data = try? Data(contentsOf: url!)
        
        //end network activity
        Menus.activityIndicator(on: false)
        
        do {
            //extract favorites data into dictionary
            var favoritesData : [String: Int] = [:]
            if let jsonData = data {
                if let json: NSDictionary = try JSONSerialization.jsonObject(with: jsonData, options: JSONSerialization.ReadingOptions.mutableContainers) as? NSDictionary {
                    if let favorites = json["favorites"] as? NSArray {
                        for item in favorites {
                            if let favorite = item as? NSDictionary {
                                let itemId = favorite["itemId"] as! String
                                let favorites = favorite["favorites"] as! Int
                                favoritesData[itemId] = favorites
                            }
                        }
                    }
                }
            }
            
            //return courses together with favorites data
            return favoritesData
        } catch {
            return [:]
        }
    }
    
    // Converts a meal and a date into a meal name param for api.
    class func getMealNameForMeal(date: Date, meal: Int) -> String {
        if date.isWeekday() {
            switch(meal) {
            case 0: return "breakfast"
            case 1: return "lunch"
            case 2: return "dinner"
            default: return ""
            }
        }
        else {
            switch(meal) {
            case 0: return "brunch"
            case 1: return "dinner"
            default: return ""
            }
        }
    }
    
    class func activityIndicator(on: Bool) {
        DispatchQueue.main.async {
            UIApplication.shared.isNetworkActivityIndicatorVisible = on
        }
    }
}
